airbyte-workers/src/main/java/io/airbyte/workers/WorkerUtils.java

```diff
@@ -53,17 +56,98 @@ public static void gentleClose(final Process process, final long timeout, final
+   *
+   * @param process - process to monitor.
+   * @param heartbeatMonitor - tracks if the heart is still beating for the given process.
+   * @param gracefulShutdownMagnitude - grace period to give the process to die after its heart stops
+   *        beating.
+   * @param gracefulShutdownTimeUnit - grace period to give the process to die after its heart stops
+   *        beating.
+   * @param checkHeartbeatMagnitude - frequency with which the heartbeat of the process is checked.
+   * @param checkHeartbeatTimeUnit - frequency with which the heartbeat of the process is checked.
+   * @param forcedShutdownMagnitude - amount of time to wait if a process needs to be destroyed
+   *        forcibly.
+   * @param forcedShutdownTimeUnit - amount of time to wait if a process needs to be destroyed
+   *        forcibly.
+   * @param forceShutdown
+   */
+  public static void gentleCloseWithHeartbeat(final Process process,
+                                              final HeartbeatMonitor heartbeatMonitor,
+                                              final long gracefulShutdownMagnitude,
+                                              final TimeUnit gracefulShutdownTimeUnit,
+                                              final long checkHeartbeatMagnitude,
+                                              final TimeUnit checkHeartbeatTimeUnit,
+                                              final long forcedShutdownMagnitude,
+                                              final TimeUnit forcedShutdownTimeUnit) {
+    gentleCloseWithHeartbeat(
+        process,
+        heartbeatMonitor,
+        gracefulShutdownMagnitude,
+        gracefulShutdownTimeUnit,
+        checkHeartbeatMagnitude,
+        checkHeartbeatTimeUnit,
+        forcedShutdownMagnitude,
+        forcedShutdownTimeUnit,
+        WorkerUtils::forceShutdown);
+  }
+
+  @VisibleForTesting
+  static void gentleCloseWithHeartbeat(final Process process,
+                                       final HeartbeatMonitor heartbeatMonitor,
+                                       final long gracefulShutdownMagnitude,
+                                       final TimeUnit gracefulShutdownTimeUnit,
+                                       final long checkHeartbeatMagnitude,
+                                       final TimeUnit checkHeartbeatTimeUnit,
+                                       final long forcedShutdownMagnitude,
+                                       final TimeUnit forcedShutdownTimeUnit,
+                                       final TriConsumer<Process, Long, TimeUnit> forceShutdown) {
+
+    while (process.isAlive() && heartbeatMonitor.isBeating()) {
       try {
-        process.waitFor(1, TimeUnit.MINUTES);
+        process.waitFor(checkHeartbeatMagnitude, checkHeartbeatTimeUnit);
       } catch (InterruptedException e) {
-        LOGGER.error("Exception while while killing the process", e);
+        LOGGER.error("Exception while while waiting for process to finish", e);
```

Comment:
As long as we're fixing terminology here, can we do something about the "while while"?